Objects/typeobject.c

```diff
@@ -2994,11 +2990,58 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
         assert(PyType_Check(base));
         dict = ((PyTypeObject *)base)->tp_dict;
         assert(dict && PyDict_Check(dict));
-        res = PyDict_GetItem(dict, name);
+        res = _PyDict_GetItem_KnownHash(dict, name, hash);
         if (res != NULL)
             break;
+        if (PyErr_Occurred()) {
+            *error = -1;
+            goto done;
+        }
```

Comment:
I dislike error handling here. I suggest to call PyObject_GetItem() if PyDict_CheckExact() is false, and to really report the error to the caller (not ignore it). Would it possible?

I don't think that it's correct to call  _PyDict_GetItem_KnownHash() if tp_dict is a dict subtype.

Reply:
Also a good catch. That complicates things a bit, though. If we have a PyUnicode subtype for the name and a non-dict mapping, we could end up calling hash() once more than needed. I've moved the hash calculation down to get a lazy calculation for the exact-dict case. This leads to a small but measurable performance degradation of about 1-2% less speedup. It's still pretty good in comparison, I think.